nav2_collision_monitor/src/velocity_polygon.cpp

```diff
@@ -0,0 +1,195 @@
+        rclcpp::ParameterValue(0.0));
+      linear_max = node->get_parameter(polygon_name_ + "." + velocity_polygon_name + ".linear_max")
+        .as_double();
+
+      // theta_min param
+      double theta_min;
+      nav2_util::declare_parameter_if_not_declared(
+        node, polygon_name_ + "." + velocity_polygon_name + ".theta_min",
+        rclcpp::ParameterValue(0.0));
+      theta_min =
+        node->get_parameter(polygon_name_ + "." + velocity_polygon_name + ".theta_min").as_double();
+
+      // theta_max param
+      double theta_max;
+      nav2_util::declare_parameter_if_not_declared(
+        node, polygon_name_ + "." + velocity_polygon_name + ".theta_max",
+        rclcpp::ParameterValue(0.0));
+      theta_max =
+        node->get_parameter(polygon_name_ + "." + velocity_polygon_name + ".theta_max").as_double();
+
+      // direction_end_angle param and direction_start_angle param
+      double direction_end_angle = 0.0;
+      double direction_start_angle = 0.0;
+      if (holonomic_) {
+        nav2_util::declare_parameter_if_not_declared(
+          node, polygon_name_ + "." + velocity_polygon_name + ".direction_end_angle",
+          rclcpp::ParameterValue(M_PI));
+        direction_end_angle =
+          node->get_parameter(polygon_name_ + "." + velocity_polygon_name + ".direction_end_angle")
+          .as_double();
+
+        nav2_util::declare_parameter_if_not_declared(
+          node, polygon_name_ + "." + velocity_polygon_name + ".direction_start_angle",
+          rclcpp::ParameterValue(-M_PI));
+        direction_start_angle =
+          node
+          ->get_parameter(polygon_name_ + "." + velocity_polygon_name + ".direction_start_angle")
+          .as_double();
+      }
+
+      SubPolygonParameter sub_polygon = {
+        poly, velocity_polygon_name, linear_min, linear_max, theta_min,
+        theta_max, direction_end_angle, direction_start_angle};
+      sub_polygons_.emplace_back(sub_polygon);
+    }
+  } catch (const std::exception & ex) {
+    RCLCPP_ERROR(
+      logger_, "[%s]: Error while getting polygon parameters: %s", polygon_name_.c_str(),
+      ex.what());
+    return false;
+  }
+  return true;
+}
+
+void VelocityPolygon::updatePolygon(const Velocity & cmd_vel_in)
+{
+  for (auto & sub_polygon : sub_polygons_) {
+    if (isInRange(cmd_vel_in, sub_polygon)) {
+      // Set the polygon that is within the speed range
+      poly_ = sub_polygon.poly_;
+
+      // Update visualization polygon
+      polygon_.polygon.points.clear();
+      for (const Point & p : poly_) {
+        geometry_msgs::msg::Point32 p_s;
+        p_s.x = p.x;
+        p_s.y = p.y;
+        // p_s.z will remain 0.0
+        polygon_.polygon.points.push_back(p_s);
+      }
+      return;
+    }
+  }
+
+  // Log for uncovered velocity
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+  RCLCPP_WARN_THROTTLE(
+    logger_, *node->get_clock(), 2.0,
```

Comment:
Store `clock_` in class definition so you don't need to lock the node every time we call this. Do `node->get_clock()` in the get params method